[visionOS] when selecting a video for docking we should select the last one that played.

jyavenard opened the pull request:
#### 1dd83ec82c6aa1fb87e980c1e5683ff86953af63
<pre>
[visionOS] when selecting a video for docking we should select the last one that played.
<a href="https://bugs.webkit.org/show_bug.cgi?id=289077">https://bugs.webkit.org/show_bug.cgi?id=289077</a>
<a href="https://rdar.apple.com/146097799">rdar://146097799</a>

Reviewed by Jer Noble.

Keep track in the VideoPresentationManagerProxy about which video was last
interacted with.
When docking, we will prefer this video when it&apos;s a child of the element in
fullscreen over the biggest.

* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::preferMediaControlsForCandidateSessionOverOtherCandidateSession):
* Source/WebCore/html/MediaElementSession.cpp:
(WebCore::MediaElementSession::mostRecentUserInteractionTime const):
(WebCore::MediaElementSession::resetPlaybackSessionState):
* Source/WebCore/html/MediaElementSession.h: Make m_mostRecentUserInteractionTime a Markable so
that we can distinguish if a media element was ever interacted with via a user gesture.
* Source/WebCore/platform/cocoa/VideoPresentationModel.h:
(WebCore::VideoPresentationModelClient::hasBeenInteractedWith):
* Source/WebCore/platform/cocoa/VideoPresentationModelVideoElement.h:
* Source/WebCore/platform/cocoa/VideoPresentationModelVideoElement.mm:
(WebCore::VideoPresentationModelVideoElement::updateForEventName):
(WebCore::VideoPresentationModelVideoElement::videoInteractedWith):
(WebCore::VideoPresentationModelVideoElement::observedEventNames):
* Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.messages.in:
* Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
(WebKit::VideoPresentationManagerProxy::hasBeenInteractedWith):
(WebKit::VideoPresentationManagerProxy::bestVideoForElementFullscreen):
* Source/WebKit/WebProcess/cocoa/VideoPresentationManager.h:
* Source/WebKit/WebProcess/cocoa/VideoPresentationManager.mm:
(WebKit::VideoPresentationInterfaceContext::hasBeenInteractedWith):
(WebKit::VideoPresentationManager::hasBeenInteractedWith):

Canonical link: <a href="https://commits.webkit.org/291939@main">https://commits.webkit.org/291939@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4be96016c8e798732ac681e795deed048e232712

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/94471 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/14063 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/3844 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/99489 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/44993 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/96521 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/14359 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/22491 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/72090 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/29411 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/97473 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/10674 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/85287 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/52422 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/10368 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/2986 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/44312 "Built successfully") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/3088 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/101534 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/21525 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/15697 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/81090 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/21773 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/81314 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/80465 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20056 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/25016 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/2389 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/14747 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/21501 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/26651 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/21184 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/24644 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/22924 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jernoble commented on Source/WebCore/platform/cocoa/VideoPresentationModelVideoElement.mm:
```diff
@@ -155,6 +155,10 @@
         }());
     }
+
+    // FIXME: We should only tag a media element as having been interacting with if those events were trigger by a user gesture.
+    if (eventName == eventNames().playEvent || eventName == eventNames().pauseEvent)
+        videoInteractedWith();
```
Can we check whether the element still has a playback restriction in place? Something like:

```
m_mediaSession->hasBehaviorRestriction(MediaElementSession::RequireUserGestureForAudioRateChange);
```

jyavenard replied:
> Can we check whether the element still has a playback restriction in place? Something like:
> 
> ```
> m_mediaSession->hasBehaviorRestriction(MediaElementSession::RequireUserGestureForAudioRateChange);
> ```

In `videoInteractedWith();` we further check that the event was fired following a user gesture.

I can further relax that if `RequireUserGestureForAudioRateChange` is false then we consider the video to have been interacted with.

jernoble commented on Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
```diff
@@ -277,6 +277,15 @@ - (BOOL)prefersStatusBarHidden
+void VideoPresentationManagerProxy::hasBeenInteractedWith(PlaybackSessionContextIdentifier contextId, bool isAudible)
+{
+    Ref model = ensureModel(contextId);
+    model->m_isAudible = isAudible;
```
It's not a great idea to reach into the model like that from within the Proxy; but also, it really seems like `isAudble` should be able to be independently updated aside from `hasBeenInteractedWith`. Maybe an event listener for `volumechange`, as well as an event listener on AudioTracks.

jernoble reviewed (approved):
r=me with nits.

jyavenard committed:
[visionOS] when selecting a video for docking we should select the last one that played.
https://bugs.webkit.org/show_bug.cgi?id=289077
rdar://146097799

Reviewed by Jer Noble.

Keep track in the VideoPresentationManagerProxy about which video was last
interacted with.
When docking, we will prefer this video when it's a child of the element in
fullscreen over the biggest.

* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::preferMediaControlsForCandidateSessionOverOtherCandidateSession):
* Source/WebCore/html/MediaElementSession.cpp:
(WebCore::MediaElementSession::mostRecentUserInteractionTime const):
(WebCore::MediaElementSession::resetPlaybackSessionState):
* Source/WebCore/html/MediaElementSession.h: Make m_mostRecentUserInteractionTime a Markable so
that we can distinguish if a media element was ever interacted with via a user gesture.
* Source/WebCore/platform/cocoa/VideoPresentationModel.h:
(WebCore::VideoPresentationModelClient::hasBeenInteractedWith):
* Source/WebCore/platform/cocoa/VideoPresentationModelVideoElement.h:
* Source/WebCore/platform/cocoa/VideoPresentationModelVideoElement.mm:
(WebCore::VideoPresentationModelVideoElement::updateForEventName):
(WebCore::VideoPresentationModelVideoElement::videoInteractedWith):
(WebCore::VideoPresentationModelVideoElement::observedEventNames):
* Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.messages.in:
* Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
(WebKit::VideoPresentationManagerProxy::hasBeenInteractedWith):
(WebKit::VideoPresentationManagerProxy::bestVideoForElementFullscreen):
* Source/WebKit/WebProcess/cocoa/VideoPresentationManager.h:
* Source/WebKit/WebProcess/cocoa/VideoPresentationManager.mm:
(WebKit::VideoPresentationInterfaceContext::hasBeenInteractedWith):
(WebKit::VideoPresentationManager::hasBeenInteractedWith):

Canonical link: https://commits.webkit.org/291939@main
